### PR TITLE
Enable Node Placeholder Pools for RFZ / readonly

### PR DESCRIPTION
**Pull request by mhoeger**
Resolves https://github.com/Azure/azure-functions-host/issues/4713

One part of this change whitelists node for placeholder pool scenarios (in RpcInitializationService).

Another part of this change makes sure to only use the placeholder pool channel on read-only apps (in WebHostLanguageWorkerChannelManager). Adding the notion of a "placeholder pool" temporarily until the one LanguageWorkerChannelManager at the webhost level has a notion of a file restart.

~~Another part of this change is to whitelist node for placeholder mode in windows. This is done so that if we turn off pools (say, if we find a bug), we won't see a significant regression in performance after having improved it.~~

~~In order to do the change above, we are checking to make sure that the major version of node.exe started in placeholder mode matches that requested by the user.~~

### CODE REVIEW

File: src/WebJobs.Script/Rpc/LanguageWorkerConstants.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@ public static class LanguageWorkerConstants
         public const int ProcessStartTimeoutSeconds = 60;
         public const string FunctionWorkerRuntimeSettingName = "FUNCTIONS_WORKER_RUNTIME";
         public const string FunctionsWorkerProcessCountSettingName = "FUNCTIONS_WORKER_PROCESS_COUNT";
+        public const string FunctionsNodeVersionSetting = "WEBSITE_NODE_DEFAULT_VERSION";
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/LanguageWorkerConstants.cs]: Move this to https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script/Environment/EnvironmentSettingNames.cs

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs [hunk 2]
```diff
@@ -116,12 +127,18 @@ internal async Task InitializeRpcServerAsync()
 
         internal Task InitializeChannelsAsync()
         {
-            if (ShouldStartInPlaceholderMode())
+            if (StartAsPlaceholderPool())
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/RpcInitializationService.cs]: instead of adding another method, move this logic inside, ShouldStartInPlaceholderMode

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 3]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
+                    // The process spun up in placeholder mode is incompatible and we should shut it down
+                    _workerChannels.TryRemove(_workerRuntime, out Dictionary<string, TaskCompletionSource<ILanguageWorkerChannel>> languageWorkerChannels);
+                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
+                }
             }
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderProcess(string selectedRuntime)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: This check is needed only for node. For java, without run from is valid

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 4]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
             _logger.LogInformation("Starting language worker channel specialization");
             _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);
+
             ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
             if (_workerRuntime != null && languageWorkerChannel != null)
             {
-                _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
-                await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                // Don't use placeholder process if app is not read-only or important environment config has changed
+                // TODO: Allow non read-only apps to use placeholder process - https://github.com/Azure/azure-functions-host/issues/4534
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Ok to remove comments. Code is self explanatory :)

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 5]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
+                if (UsePlaceholderProcess(_workerRuntime))
+                {
+                    _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: change to log debug. Add logging here, valid placeholdermode process and also in the else block, incompatible palceholdermode process,

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 6]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
+            {
+                return false;
+            }
+
+            // App must have specific environment configurations match
+            if (_initialEnvironmentConfig.TryGetValue(selectedRuntime, out Dictionary<string, string> environmentConfig))
+            {
+                foreach (string settingKey in environmentConfig.Keys)
+                {
+                    var currentValue = _environment.GetEnvironmentVariable(settingKey);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: tested this with ~10, 10.14.0 and ~8 , 8.0.1?

File: test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs [hunk 7]
```diff
@@ -59,7 +59,7 @@ public WebHostLanguageWorkerChannelManagerTests()
             _languageWorkerProcessFactory = new Mock<ILanguageWorkerProcessFactory>();
             _languageWorkerProcessFactory.Setup(m => m.CreateLanguageWorkerProcess(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<string>())).Returns(_languageWorkerProcess.Object);
 
-            _languageWorkerChannelFactory = new TestLanguageWorkerChannelFactory(_eventManager, null, _scriptRootPath);
+            _languageWorkerChannelFactory = new TestLanguageWorkerChannelFactory(_eventManager, new TestLogger("WebHostLanguageWorkerChannelManagerTests"), _scriptRootPath);
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs]: move creating testLogger to constructor. Are you using this to verify logs? if not undo

**Comment by pragnagopa** [comment]
If this change is scoped to placeholder pools, no need to check for node version as poolid matching logic takes care of it

File: src/WebJobs.Script/Rpc/LanguageWorkerConstants.cs (hunk 1 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/LanguageWorkerConstants.cs]: After looking through the PR and offline conversation, it does make sense to enable placeholder just for placepholder pools. Please go ahead and remove version matching check. PlaceholderPoolId matching logic maps any major 8 version to ~8 and any major 10 version to ~10.

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 8]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
+                // TODO: Allow non read-only apps to use placeholder process - https://github.com/Azure/azure-functions-host/issues/4534
+                if (UsePlaceholderProcess(_workerRuntime))
+                {
+                    _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
+                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                }
+                else
+                {
+                    // The process spun up in placeholder mode is incompatible and we should shut it down
+                    _workerChannels.TryRemove(_workerRuntime, out Dictionary<string, TaskCompletionSource<ILanguageWorkerChannel>> languageWorkerChannels);
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: >_workerChannels.TryRemove(_workerRuntime, out Dictionary<string, TaskCompletionSource<ILanguageWorkerChannel>> languageWorkerChannels); [](start = 20, length = 135)

_workerChannels.TryRemove(_workerRuntime, out _ ) ?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 9]
```diff
@@ -102,16 +109,54 @@ public async Task SpecializeAsync()
...
+            // App must be readonly
+            if (!_environment.FileSystemIsReadOnly())
+            {
+                return false;
+            }
+
+            // App must have specific environment configurations match
+            if (_initialEnvironmentConfig.TryGetValue(selectedRuntime, out Dictionary<string, string> environmentConfig))
+            {
+                foreach (string settingKey in environmentConfig.Keys)
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: >foreach (string settingKey in environmentConfig.Keys) [](start = 16, length = 53)

If you're looping through the keys anyway, does this need to be a dictionary? Can it just be a List<> instead?

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs (hunk 2 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/RpcInitializationService.cs]: @pragnagopa -  I think we want to keep it as separate logic because we'd only like to start the relevant specific process?

File: test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs (hunk 7 above)
Comment (mhoeger) [reply test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs]: I think we need to do this to, otherwise tests fail with a null exception here! In the constructor of CreateTestLanguage, it takes in an "IMetricsLogger" and not a "TestLogger"

Comment (pragnagopa) [reply test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs]: Add tests that verify specialization logs, see here for example on verifying test logs: azure-functions-host\test\WebJobs.Script.Tests\Rpc\LanguageWorkerChannelTests.cs

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 10]
```diff
@@ -102,11 +102,23 @@ public async Task SpecializeAsync()
...
+                if (_environment.FileSystemIsReadOnly())
+                {
+                    _logger.LogDebug("Loading environment variables for runtime: {runtime}", _workerRuntime);
+                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                }
+                else
+                {
+                    _logger.LogDebug("Shutting down placeholder worker. Worker is not compatible for runtime: {runtime}", _workerRuntime);
+                    // If we need to allow file edits, we should shutdown the webhost channel on specialization.
+                    _workerChannels.TryRemove(_workerRuntime, out _);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: delete: ` _workerChannels.TryRemove(_workerRuntime, out _);` as ShutdownChannelIfExistsAsync removes and then disposes channel

Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: good catch, done!

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 11]
```diff
@@ -102,11 +102,22 @@ public async Task SpecializeAsync()
...
+
             ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
             if (_workerRuntime != null && languageWorkerChannel != null)
             {
-                _logger.LogInformation("Loading environment variables for runtime: {runtime}", _workerRuntime);
-                await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
+                // TODO: https://github.com/Azure/azure-functions-host/issues/4534 Don't kill non read-only processes once linked issue is resolved
+                if (_environment.FileSystemIsReadOnly())
+                {
+                    _logger.LogDebug("Loading environment variables for runtime: {runtime}", _workerRuntime);
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: this check is only needed for node.

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 12]
```diff
@@ -104,25 +104,33 @@ public async Task SpecializeAsync()
...
+                await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
             }
+
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderChannel(string workerRuntime, bool channelExists)
+        {
+            if (!string.IsNullOrEmpty(workerRuntime) && channelExists)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: can you please change it to

```
 public async Task SpecializeAsync()
        {
            _logger.LogInformation("Starting language worker channel specialization");
            _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);

            ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
            if (languageWorkerChannel != null)
            {
                if (UsePlaceholderChannel(_workerRuntime))
                {
                    _logger.LogDebug("Loading environment variables for runtime: {runtime}", _workerRuntime);
                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
                }
                else
                {
                    _logger.LogDebug("Shutting down placeholder worker. Worker {workerId }is not compatible for runtime: {runtime}", _workerRuntime, languageWorkerChannel.Id);
                    // If we need to allow file edits, we should shutdown the webhost channel on specialization.
                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
                }
            }

            _shutdownStandbyWorkerChannels();
            _logger.LogDebug("Completed language worker channel specialization");
        }

        internal bool UsePlaceholderChannel(string workerRuntime)
        {
            if (!string.IsNullOrEmpty(_workerRuntime) && string.Equals(_workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
            {
                return _environment.FileSystemIsReadOnly();
            }
            return true;
        }
```

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 13]
```diff
@@ -104,25 +104,33 @@ public async Task SpecializeAsync()
...
+                _logger.LogDebug("Shutting down placeholder worker. Worker is not compatible for runtime: {runtime}", _workerRuntime);
+                // If we need to allow file edits, we should shutdown the webhost channel on specialization.
+                await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
             }
+
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderChannel(string workerRuntime, bool channelExists)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: add unit tests just for UsePlaceholderChannel. Simpler to verify conditions

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 12 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: done!

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 13 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: I think I did in WebHostLanguageWorkerManagerTests, is that what you're thinking of too?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 14]
```diff
@@ -102,16 +102,38 @@ public async Task SpecializeAsync()
...
+                }
             }
+
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderChannel(string workerRuntime)
+        {
+            if (!string.IsNullOrEmpty(workerRuntime))
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: use as it is easier to read, if node check for run from zip
```
 internal bool UsePlaceholderChannel(string workerRuntime)
        {
            if (!string.IsNullOrEmpty(_workerRuntime) && string.Equals(_workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
            {
                return _environment.FileSystemIsReadOnly();
            }
            return true;
        }
```

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 13 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: just tests that verify true or false when calling UsePlaceholderChannel

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 14 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: i this would fail for the case when _workerRuntime is null or empty, but we would return "true" for UsePlaceholderChannel. how does this look for being more readable? 

```
            if (!string.IsNullOrEmpty(workerRuntime))
            {
                // Special case: node apps must be read-only to use the placeholder mode channel
                // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
                if (string.Equals(_workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
                {
                    return _environment.FileSystemIsReadOnly();
                }
                return true;
            }
            return false;
```

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Looks good! yea please make the change

**Comment by pragnagopa** [comment]
Please test on private stamp:
- Node with and without zip
- Java with and with out zip
- C#

File: test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs [hunk 15]
```diff
@@ -146,6 +148,84 @@ public async Task ShutdownStandyChannels_WorkerRuntime_Node_Set()
...
+            var initializedChannel = await _languageWorkerChannelManager.GetChannelAsync(LanguageWorkerConstants.NodeLanguageWorkerName);
+            Assert.Null(initializedChannel);
+        }
+
+        [Theory]
+        [InlineData("node", "1", true)]
+        [InlineData("node", "0", false)]
+        [InlineData("java", "1", true)]
+        [InlineData("java", "0", true)]
+        public void TryParseStatusCode_ReturnsExpectedResult(string workerRuntime, string runFromZip, bool expectedReturn)
```
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs]: rename to test to UsePlaceholderProcess_ReturnsExpectedResult

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs [hunk 16]
```diff
@@ -116,12 +122,18 @@ internal async Task InitializeRpcServerAsync()
...
+            {
+                return _webHostlanguageWorkerChannelManager.InitializeChannelAsync(_workerRuntime);
+            }
+            else if (ShouldStartInPlaceholderMode())
             {
                 return InitializePlaceholderChannelsAsync();
             }
-
-            return InitializeWebHostRuntimeChannelsAsync();
+            else
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/RpcInitializationService.cs]: remove last else block as that would be the default

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 17]
```diff
@@ -102,16 +102,42 @@ public async Task SpecializeAsync()
         {
             _logger.LogInformation("Starting language worker channel specialization");
             _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);
+
             ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
-            if (_workerRuntime != null && languageWorkerChannel != null)
+            if (languageWorkerChannel != null)
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: change to  otherwise, shutdownIfChannelExists will throw null ref. We cannot shutdown placeholder channel here if worker runtime not set. FunctionDispatcher will take care of it.
```
public async Task SpecializeAsync()
        {
            _logger.LogInformation("Starting language worker channel specialization");
            _workerRuntime = _environment.GetEnvironmentVariable(LanguageWorkerConstants.FunctionWorkerRuntimeSettingName);

            ILanguageWorkerChannel languageWorkerChannel = await GetChannelAsync(_workerRuntime);
            if (_workerRuntime != null && languageWorkerChannel != null)
            {
                if (UsePlaceholderChannel(_workerRuntime))
                {
                    _logger.LogDebug("Loading environment variables for runtime: {runtime}", _workerRuntime);
                    await languageWorkerChannel.SendFunctionEnvironmentReloadRequest();
                }
                else
                {
                    _logger.LogDebug("Shutting down placeholder worker. Worker is not compatible for runtime: {runtime}", _workerRuntime);
                    // If we need to allow file edits, we should shutdown the webhost channel on specialization.
                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
                }
            }

            _shutdownStandbyWorkerChannels();
            _logger.LogDebug("Completed language worker channel specialization");
        }

        internal bool UsePlaceholderChannel(string workerRuntime)
        {
            if (!string.IsNullOrEmpty(workerRuntime))
            {
                // Special case: node apps must be read-only to use the placeholder mode channel
                // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
                if (string.Equals(workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
                {
                    return _environment.FileSystemIsReadOnly();
                }
            }
            return true;
        }
```

Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Ok! Should we also not make ShutdownChannelIfExists not throw? In my mind, the "IfExists" implies that this method should only try to shutdown something if it's there but otherwise continue executing safely.

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: calling shutdown when workerRuntime is null is a no-op and is intentional to throw error as it enforces not calling this method with null. ShutdownChannelIfExists means that, for a valid runtime, if the list contains a channel for that runtime shut it down.

**Comment by mhoeger** [comment]
@pragnagopa - addressed CR comments! Could I get a final review? I've tested Node and Java with RFZ but won't merge until i've tested Node/Java without RFZ and C# with/without RFZ.

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 18]
```diff
@@ -102,16 +102,42 @@ public async Task SpecializeAsync()
...
+                    await ShutdownChannelIfExistsAsync(_workerRuntime, languageWorkerChannel.Id);
+                }
             }
+
             _shutdownStandbyWorkerChannels();
             _logger.LogDebug("Completed language worker channel specialization");
         }
 
+        internal bool UsePlaceholderChannel(string workerRuntime)
+        {
```
Comment (pragnagopa) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: simplify - this method needs to return true if it is ever called with runtime not set. FunctionDispatcher will  shutdown placeholder channels after jobhost initializes

```
internal bool UsePlaceholderChannel(string workerRuntime)
        {
            if (!string.IsNullOrEmpty(workerRuntime))
            {
                // Special case: node apps must be read-only to use the placeholder mode channel
                // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
                if (string.Equals(workerRuntime, LanguageWorkerConstants.NodeLanguageWorkerName))
                {
                    return _environment.FileSystemIsReadOnly();
                }
            }
            return true;
        }
```

File: test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs (hunk 15 above)
Comment (pragnagopa) [reply test/WebJobs.Script.Tests/Rpc/WebHostLanguageWorkerChannelManagerTests.cs]: missed this?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 18 above)
Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: If workerRuntime isn't set, then we won't be able to use the channel that was initialized in placeholder mode because we won't know which channel we should be using (ex: to specialize/sending environment variables to). So I think that if workerRuntime is not set, then we should not use any channel that was started in placeholder mode (and if we were to support workerRuntime that is not set, there are more changes needed to support that, like sending environment variables to all worker channels started in placeholder mode). 

The check for null or empty workerRuntime isn't necessary either because before it is called, we make sure that workerRuntime is not null or empty. I want to keep that logic there though in case some other method chooses to call `UsePlaceholderChannel` without the same check, and I think in this case the result should be "false"

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: unfortunately, if workerruntime is not set, there is no action at webhost level. Placeholder channel needs to be disposed at jobhost level. To avoid others calling into this method, make this method private to webhostchannelmanager?

Comment (mhoeger) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Sounds good! I think I made it internal for unit tests, but we have tests on SpecializeAsync too so I can make this private or just put it into the logic for SpecializeAsync.

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: yes either one works! thanks!
